### ISSUE TITLE
[Feature Request] CanoniconalizedSet<C, E>

### ISSUE THREAD

**Issue by seanocali** [dart-lang/language#2607]
Would love to be able to use a case-insensitive HashSet of String.

**Comment by Wdestroier** [comment]
All you have to do is implement a HashSet and call `key.toLowerCase()` before each operation. This issue should be moved to the dart-lang/sdk repository or could be asked on StackOverflow.

**Comment by TimWhiting** [comment]
You could use this map as a set: https://pub.dev/documentation/collection/latest/collection/CanonicalizedMap-class.html

**Comment by seanocali** [comment]
> You could use this map as a set: https://pub.dev/documentation/collection/latest/collection/CanonicalizedMap-class.html

Yes that's what I'm currently doing. Just simply requesting a Set version of this.

**Comment by seanocali** [comment]
> All you have to do is implement a HashSet and call `key.toLowerCase()` before each operation. This issue should be moved to the dart-lang/sdk repository or could be asked on StackOverflow.

Wouldn't you have to iterate it in order to do that, thus defeating the purpose of a hash table.

**Comment by Cat-sushi** [comment]
> Yes that's what I'm currently doing. Just simply requesting a Set version of this.

How can you canonicalize objects with `Set` version?

**Comment by seanocali** [comment]
> > Yes that's what I'm currently doing. Just simply requesting a Set version of this.
> 
> How can you canonicalize objects with `Set` version?

The same way it's done with the Map.

**Comment by seanocali** [comment]
This should probably be moved to: [https://github.com/dart-lang/collection/issues](https://github.com/dart-lang/collection/issues)

**Comment by Cat-sushi** [comment]
> The same way it's done with the Map.

What is your use case?

**Comment by lrhn** [comment]
If you want a set of canonicalized values, I'd just canonicalize the values before putting them into the set.
I don't see a set-which-canonicalizes-on-add as a big enough value increment over that to be worth a separate class.

If you want a set which considers some, otherwise not equal, objects equal because they canonicalize to the same thing, but still want to keep the uncanonicalized values, you can use `HashSet` with a custom equality.

If you want that *and* keeping the canonicalized version of the value around, so you don't need to re-canonicalize on every `equals` call, I'd just a map instead, and canonicalize the keys. The `CanonicalizedMap` mentioned above can help with that.
